flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/MergeTableAsUtil.java

```diff
@@ -0,0 +1,322 @@
+import org.apache.flink.table.planner.calcite.FlinkTypeFactory;
+import org.apache.flink.table.types.logical.LogicalType;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.SqlDataTypeSpec;
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.validate.SqlValidator;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static org.apache.flink.table.planner.calcite.FlinkTypeFactory.toLogicalType;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeCasts.supportsImplicitCast;
+import static org.apache.flink.table.types.utils.TypeConversions.fromLogicalToDataType;
+
+/** A utility class with logic for handling the {@code CREATE TABLE ... AS SELECT} clause. */
+public class MergeTableAsUtil {
+    private final SqlValidator validator;
+    private final Function<SqlNode, String> escapeExpression;
+    private final DataTypeFactory dataTypeFactory;
+
+    MergeTableAsUtil(
+            SqlValidator validator,
+            Function<SqlNode, String> escapeExpression,
+            DataTypeFactory dataTypeFactory) {
+        this.validator = validator;
+        this.escapeExpression = escapeExpression;
+        this.dataTypeFactory = dataTypeFactory;
+    }
+
+    /**
+     * Merges the schema part of the {@code sqlCreateTableAs} with the {@code sourceSchema}.
+     *
+     * <p>It is expected that the {@code sourceSchema} contains only physical/regular columns, which
+     * is behavior of the CTAS statement to generate such schema.
+     *
+     * <p>Columns of {@code sourceSchema} are appended to the schema of {@code sqlCreateTableAs}. If
+     * a column in the {@code sqlCreateTableAs} is already defined in {@code sourceSchema}, then the
+     * types of the columns are implicit cast and must be compatible based on the implicit cast
+     * rules. If they're compatible, then the column position in the schema stays the same as
+     * defined in the appended {@code sourceSchema}.
+     */
+    public Schema mergeSchemas(SqlCreateTableAs sqlCreateTableAs, ResolvedSchema sourceSchema) {
+        SchemaBuilder schemaBuilder =
+                new SchemaBuilder(
+                        (FlinkTypeFactory) validator.getTypeFactory(),
+                        dataTypeFactory,
+                        validator,
+                        escapeExpression);
+
+        schemaBuilder.mergeColumns(
+                sqlCreateTableAs.getColumnList(),
+                Schema.newBuilder().fromResolvedSchema(sourceSchema).build().getColumns());
+
+        return schemaBuilder.build();
+    }
+
+    /**
+     * Builder class for constructing a {@link Schema} based on the rules of the {@code CREATE TABLE
+     * ... AS SELECT} statement.
+     */
+    private static class SchemaBuilder {
```

Comment:
The way they're merged are different. The LIKE util adds the `sourceSchema` at the beginning and the sink columns at the end. It also depends on the LIKE STRATEGIES on how primary/partition keys, watermarks are merged, either OVERWRITING, IGNORING, EXCLUDING. There's no strategy for schema columns yet. 

I was tempted to re-use that util class and add the support for column overwriting (currently, the LIKE util throws if a sink column exists in the source column), but I thought I will start messing with the CREATE TABLE LIKE and its behavior. I instead decided to create this other utility to prevent changing logic of another statement.